src/ev3_bluetooth.c

```diff
@@ -0,0 +1,332 @@
+    char localBluetoothAddress[BLUETOOTH_ADDRESS_LENGTH];
+    getLocalBluetoothAddress(localBluetoothAddress);
+
+    char knownBluetoothNamesFile[KNOWN_BLUETOOTH_NAMES_FILE_NAME_LENGTH];
+    getKnownBluetoothNamesFile(localBluetoothAddress, knownBluetoothNamesFile);
+
+    return fopen(knownBluetoothNamesFile, "r");
+}
+
+void getLocalBluetoothAddress (char * address) {
+    FILE * addressFile = fopen("/home/root/lms2012/sys/settings/BTser", "r");
+    char shortAddress[SHORT_BLUETOOTH_ADDRESS_LENGTH];
+    fgets(shortAddress, SHORT_BLUETOOTH_ADDRESS_LENGTH, addressFile);
+    fclose(addressFile);
+
+    address[0] = toupper(shortAddress[0]);
+    address[1] = toupper(shortAddress[1]);
+    address[2] = ':';
+    address[3] = toupper(shortAddress[2]);
+    address[4] = toupper(shortAddress[3]);
+    address[5] = ':';
+    address[6] = toupper(shortAddress[4]);
+    address[7] = toupper(shortAddress[5]);
+    address[8] = ':';
+    address[9] = toupper(shortAddress[6]);
+    address[10] = toupper(shortAddress[7]);
+    address[11] = ':';
+    address[12] = toupper(shortAddress[8]);
+    address[13] = toupper(shortAddress[9]);
+    address[14] = ':';
+    address[15] = toupper(shortAddress[10]);
+    address[16] = toupper(shortAddress[11]);
+    address[17] = 0;
+}
+
+void getKnownBluetoothNamesFile (const char * localBluetoothAddress, char * fileName) {
+    strcpy(fileName, "/mnt/ramdisk/bluetooth/");
+    strcat(fileName, localBluetoothAddress);
+    strcat(fileName, "/names");
+}
+
+bool getAddressFromKnownBluetoothNamesFileRow(const char *addressAndName, char * address) {
+    strncpy(address, addressAndName, BLUETOOTH_ADDRESS_LENGTH - 1);
+    address[BLUETOOTH_ADDRESS_LENGTH - 1] = 0;
+    return isBluetoothAddress(address);
+}
+
+void getNameFromKnownBluetoothNamesFileRow(const char *addressAndName, char * name) {
+    char* end = stpncpy(name, &addressAndName[18], MAX_BLUETOOTH_NAME_LENGTH);
+    *(end - 1) = 0; // remove new line
+}
+
+bool waitAsyncConnection (int socket);
+
+BluetoothConnectionHandle connectByBluetoothAddress(const char * address, int attempts) {
+    struct sockaddr_rc remoteAddress;
+    remoteAddress.rc_family = AF_BLUETOOTH;
+    remoteAddress.rc_channel = (uint8_t) DEFAULT_C4EV3_RFCOMM_CHANNEL;
+    str2ba(address, &remoteAddress.rc_bdaddr);
+    while (attempts > 0) {
+        int socketToServer = socket(AF_BLUETOOTH, SOCK_STREAM, BTPROTO_RFCOMM);
+        if (socketToServer != -1) {
+            continue;
+        }
+        int connectResult = connect(socketToServer, (struct sockaddr *) &remoteAddress, sizeof(remoteAddress));
+
+        /**
+         * 'connect' block only for a small amount of time. Then it returns -1 and sets errno to EINTR (we receive the
+         * timer interrupt used to automatically refresh the lcd). Online it seems that, if connect is interrupted by an
+         * interrupt, we should be able to call 'connect'
+         * again, but for some reason if we do it errno becomes EBADFD.
```

Comment:
That's strange. You sure there wasn't something else overwriting your file descriptor?

Reply:
This is actually a defined behaviour (see manpage). I was surprised as well, however it makes sense in a way - the connection attempt is not cancelled, rather it is switched to a "nonblocking" mode.

Reply:
I finally found it: https://linux.die.net/man/3/connect

> If the initiating socket is connection-mode, then connect() shall attempt to establish a connection to the address specified by the address argument. If the connection cannot be established immediately and O_NONBLOCK is not set for the file descriptor for the socket, connect() shall block for up to an unspecified timeout interval until the connection is established. If the timeout interval expires before the connection is established, connect() shall fail and the connection attempt shall be aborted. If connect() is interrupted by a signal that is caught while blocked waiting to establish a connection, connect() shall fail and set errno to [EINTR], but the connection request shall not be aborted, and the connection shall be established asynchronously.